parser-parse.go

```diff
@@ -113,17 +114,57 @@ func (ps *Parser) ParseDocument(doc *html.Node, pageURL *nurl.URL) (Article, err
 	validByline := strings.ToValidUTF8(finalByline, "")
 	validExcerpt := strings.ToValidUTF8(excerpt, "")
 
+	publishedTime := ps.getDate(metadata, "publishedTime")
+	modifiedTime := ps.getDate(metadata, "modifiedTime")
+
 	return Article{
-		Title:       validTitle,
-		Byline:      validByline,
-		Node:        readableNode,
-		Content:     finalHTMLContent,
-		TextContent: finalTextContent,
-		Length:      charCount(finalTextContent),
-		Excerpt:     validExcerpt,
-		SiteName:    metadata["siteName"],
-		Image:       metadata["image"],
-		Favicon:     metadata["favicon"],
-		Language:    ps.articleLang,
+		Title:         validTitle,
+		Byline:        validByline,
+		Node:          readableNode,
+		Content:       finalHTMLContent,
+		TextContent:   finalTextContent,
+		Length:        charCount(finalTextContent),
+		Excerpt:       validExcerpt,
+		SiteName:      metadata["siteName"],
+		Image:         metadata["image"],
+		Favicon:       metadata["favicon"],
+		Language:      ps.articleLang,
+		PublishedTime: publishedTime,
+		ModifiedTime:  modifiedTime,
 	}, nil
 }
+
+// getDate tries to get a date from metadata, and parse it using a list of known formats.
+func (ps *Parser) getDate(metadata map[string]string, fieldName string) *time.Time {
+	dateStr, ok := metadata[fieldName]
+	if ok && len(dateStr) > 0 {
+		return getParsedDate(dateStr)
+	}
+	return nil
+}
+
+// getParsedDate tries to parse a date string using a list of known formats.
+// If the date string can't be parsed, it will return nil.
+func getParsedDate(dateStr string) *time.Time {
+	// Following formats have been seen in the wild.
+	formats := []string{
```

Comment:
Some of these are defined under `time` constants, isn't better to use those?

Reply:
Some are defined as `time` constants, but the reason why I do not use the `time.RFCxxx` is that we essentially need the longest ones first since the loop breaks at the first correct match. Using the constants makes it very difficult to know the length/precision of the date/time pattern, so I used explicit strings instead.

It's more for convenience – both can be done obviously :)

Reply:
That totally makes sense, let's leave it like it is 👍